### PR TITLE
[PoC] Allow broadcasting keys to all panes in tab

### PR DESCRIPTION
**Pull request by Don-Vito**
## Summary of the Pull Request
Introduces ToggleInputBroadcast command,
that makes Terminal Tab to broadcast its input to all panes.

This is an early draft to see if the implementation
and UX make any sense at all.

## References
This is an attempt to address #2634.

## PR Checklist
* [x] CLA signed. 
* [ ] Tests added/passed
* [ ] Documentation updated. 
* [ ] Schema updated.
* [ ] I've discussed this with core contributors already. 

## Detailed Description of the Pull Request / Additional comments
* Store the IsInputBroadcastActive flag on the tab level
* Add KeySent and CharSent events to TermControl
* Register on this events in Tab and send the keys / chars
to other terminals if IsInputBroadcastActive 

One of the dilemmas was were to hook the input.
The most robust way would be to do it upon SendInputToConnection.
But it will prevent auto scrolling and deselection.

TODO
* [ ] Fix the cursor blinking
* [ ] Use accent colors for color coding

### CODE REVIEW

**Comment by Don-Vito** [comment]
InputBroadcast

**Comment by Don-Vito** [comment]
@zadjii-msft, @carlos-zamora - this is a PoC. Simply want to hear some feedback to learn if this is a good direction (product and engineering wise).

**Comment by carlos-zamora** [comment]
I like this proof of concept. I think it'd be nice if the focus border on the pane would encompass all of the panes somehow, just to make it more clear that you're in broadcast mode.

I'm not familiar with how other terminals do this UI-wise. Might be worth taking a look there.

For anybody that comes along, here's a closer look at the new tab icon:
Broadcast icon

I'll take a closer look at the actual code changes later, but I'm liking the direction this is going in 😀

**Comment by zadjii-msft** [comment]
Alright so 👀 wow, this is awesome.

I think my real concern though is how the plumbing for this works. It seems like you've got it set up so that input can be broadcast to all the panes in a tab, but not a _subset of the panes in a tab_, or to panes in other tabs. I think that's going to end up being one of the more important parts of this functionality. The actions that were in the OP look like they are:

* **A** Only send input to the active pane
* **B** send input to all panes in all tabs
* **C** send input to all panes in the current tab
* **D** toggle sending input to the current pane

So I'm not sure exactly how those play together in certain scenarios, and unfortunately don't have a Mac to test iterm2 myself. The scenario that gets me is like:
1. Enable broadcast input for a pane in tab 1
2. switch to tab 2
3. execute action C to send input to all the panes in the current tab

Does that mean that you're now broadcasting to the pane in tab 1 and all the panes in tab 2? If at this point you:

4. Switch to tab 3

now who's getting input? All the panes in tab 3? All the panes in tab 2? All the panes in 2, plus that first pane, plus the active pane in 3?

<hr>

Okay maybe I'm off track. This is a good proof-of-concept, and the bubbling of the events seems sound. So I guess my concern is in the trick of how do we determine which panes should be getting the broadcasted input. I'm _all for_ incremental progress, but I want to make sure we have our eyes on the end goal. 

Maybe we should have a plan for what the actions will look like in the end, so we can start here and work towards those actions? I'd hate to have release notes with "added the `toggleBroadcastInput` action" followed immediately by "removed the `toggleBroadcastInput` action, which has been replaced with `whateverThingWeComeUpWith`"

**Comment by Don-Vito** [comment]
@zadjii-msft  - I have a subset of what you mentioned, due to the read-only tabs :smile:

Please see the following gif below. I tried btw to use color coding of the boarder to show what panes are read-only and what get broadcast (but I guess this gonna be an accessibility issue we can resolve only with pane headers)

InputBroadcast2

**Comment by zadjii-msft** [comment]
I kinda like the border colorizing. I agree that pane headers would be the _best_ way of solving the indicator problem. However, those are probably still a good ways out. The colorized borders might be a decent temporary solution. I'd definitely want the other broadcasted panes identified in _some way_. 

I wonder if we should be using some variant of the accent color ([link](https://docs.microsoft.com/en-us/windows/uwp/design/style/color#accent-color-palette)) for the broadcasted pane's borders. At least until there's a proper way to stylize the borders with theming (#3327). Using a variation on the accent color would make sure that the color of a broadcasted pane border is always distinct, even if the user has some green-like color set as their accent color.

> The next thing we can do is to add "exclude from broadcast" instead of readonly

☝️ I like that. After that we can work out how we broadcast to panes in other tabs too.

<hr>

Not to do the whole spec in comments, but I'm thinking:

We'll maintain a set of panes that are the "broadcast set". All the panes in the broadcast set would also get the KeySent and CharSent events (in addition to the active pane, which can be a part of the broadcast set). If a pane is read-only in the broadcast set, then it _won't_ handle those broadcasted events (obviously). 

As far as actions, we're looking at something like:

* **A** Only send input to the active pane
  * Remove all the panes from the broadcast set
* **B** send input to all panes in all tabs
  * If all the panes are in the broadcast set, remove them all. Otherwise, add all panes in all tabs to the broadcast set.
* **C** send input to all panes in the current tab
  * If all the panes in the current tab are in the broadcast set, remove them from the broadcast set. Otherwise, add all the panes from this tab to the broadcast set.
* **D** toggle sending input to the current pane
  * If this pane is in the broadcast set, remove it. Otherwise add it.

I'm gonna try socializing this with the team today, but we've already got a lot on the agenda today 😬 this might get bumped slowly, sorry in advance

**Comment by Don-Vito** [comment]
Thanks! Don't worry about the timing - my last weeks are hectic, so I don't really get to code 😢 
I hope I won't forget what I've written here :smile:

**Comment by Don-Vito** [comment]
@zadjii-msft - just to make sure, given that we already have the read-only, can we postpone the exclusion to the next phases?

**Comment by zadjii-msft** [comment]
@Don-Vito Sorry, I hadn't had a chance to come back around on this since my last post. I think I want to settle the design of this a _bit_ more. What you've got here is basically the  "**C** send input to all panes in the current tab" action. So I think at this point I just want to try and come up with a design for the actual serialization of all these various actions. That way we can align this PR as implementation for the scenario **C** action. We won't need to worry about the rest of the implementation now, but I want to have the design be future proof. If anything, the only thing that will need to change for _this PR_ is the json key name. 

I think I've got a bit of time open in my schedule later this week to whip something more formal up. I'll put together 25% of a spec and socialize that with the team.

**Comment by zljubisic** [comment]
In Linux terminator you can define groups, and than put each pane in one of defined groups.
Afterwards, you can choose broadcasting to all panes, only certain group or no broadcast at all.
Anyway, what you have done so far looks quite promissing. 👍

**Comment by zadjii-msft** [comment]
Okay sorry for the radio silence here. We talked this over as a team - not _all_ the details in #9365, but we did talk over the important conclusion. Regardless how we end up going about implementing the rest of the broadcast input functionality, we're definitely cool with `{ "command": "toggleBroadcastInput" }` (without other args) meaning "send input to all the panes in this tab". We can sort out the rest of the details in the bigger spec review, but this is a good enough place to start ☺️ 

I'll go take a closer look at the code later. Thanks again for your patience!

**Comment by Don-Vito** [comment]
@zadjii-msft - please pay attention that  the solution is not final (e.g., I still need to make sure that the cursor blinking is explicitly enabled and disabled in all panes)! If you believe that this PR is a good start, I will continue investing.

**Comment by zadjii-msft** [comment]
Yea, I definitely think that this PR is a good enough start. Hopefully by the time we're done reviewing it, we can sign off on #9365 and have follow-up issues ready to file ☺️

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 1]
```diff
@@ -1111,6 +1122,25 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
         }
     }
 
+    // Method Description:
+    // - Invokes TrySendKeyEvent and triggers KeySent event
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControl.cpp:1126]:
```suggestion
    // - Invokes TrySendKeyEvent and triggers KeySent event. This is only called
    //   for keys that are directly typed into this control
```

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 2]
```diff
@@ -1111,6 +1122,25 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
+    // - keyDown: If true, the key was pressed, otherwise the key was released.
+    bool TermControl::_TrySendKeyEvent(const WORD vkey,
+                                       const WORD scanCode,
+                                       const ControlKeyStates keyStates,
+                                       const bool keyDown)
+    {
+        const auto modifiers = keyStates.Value();
+        const auto result = TrySendKeyEvent(vkey, scanCode, modifiers, keyDown);
+        auto keySentArgs = winrt::make<KeySentEventArgs>(vkey, scanCode, modifiers, keyDown);
+        _keySentHandlers(*this, std::move(keySentArgs));
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControl.cpp:1140]:
```suggestion
        // Raise the keySent event to broadcast this key to other potential terminal controls
        _keySentHandlers(*this, std::move(keySentArgs));
```

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
I've got a couple spots I'm suggesting comments only because it's not immediately obvious the difference between `_TrySendKeyEvent` and `TrySendKeyEvent`. Plus I've got proposal for the pane border color, but I'd love more input on that. 

In general, this seems pretty straightforward.

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp [hunk 3]
```diff
@@ -53,6 +53,7 @@ static constexpr std::string_view BreakIntoDebuggerKey{ "breakIntoDebugger" };
 static constexpr std::string_view FindMatchKey{ "findMatch" };
 static constexpr std::string_view TogglePaneReadOnlyKey{ "toggleReadOnlyMode" };
 static constexpr std::string_view NewWindowKey{ "newWindow" };
+static constexpr std::string_view ToggleInputBroadcastKey{ "toggleInputBroadcast" };
```
[the remark this answers is not in the document]
Comment (WSLUser) [reply src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:56]: I prefer the latter option personally. Sounds more sensible.

**Comment by zadjii-msft** [comment]
_yea jeez bot cool it_. Take your time - we're in no huge rush to merge this one ☺️

**Comment by zljubisic** [comment]
Do you have a prerelease version or something like that in order to check it?

**Comment by zadjii-msft** [comment]
Hey so for the record, I'm about to blow up `TermControl` in #9820. Merging this with that is probably gonna be a huge pain, lemme know if you need any help (whenever you come back around on this one)

**Comment by Don-Vito** [comment]
> Hey so for the record, I'm about to blow up `TermControl` in #9820. Merging this with that is probably gonna be a huge pain, lemme know if you need any help (whenever you come back around on this one)

NP. I think this one can be done in parts. We can start with:
1. Indication of read-only panes. This will create a platform for pane status indication.
2. Indication of panes getting input (both coloring and cursor blinking)
3. Broadcast support in term control
4. Tab level broadcasting
5...

We can start with something very basic and improve it gradually. Because, it seems that smaller contributions work while in bigger ones it takes time and by reply I already lose the focus. 

@zadjii-msft - do you want to go through this journey with me?

**Comment by zadjii-msft** [comment]
@Don-Vito I'm happy to help with this. It generally sems like a good plan, and I think it'll make a lot of folks happy. The only thing that I'm not totally sure about is 

> Indication of read-only panes. This will create a platform for pane status indication.

That one I'm a little worried about only for a couple reasons. One - if we just go with like, a red border, then I'm worried that someone will have their accent color set to that same color, so the border doesn't actually provide any new info. However, if we don't go with the border color, then what _do_ we go with? That question starts getting into the UI design realm, and those kinds of questions are always harder to come to a consensus. We've already got so many design questions swirling around the settings UI, I'm worried that this would more easily get lost in those weeds.

So I kinda want to skip that particular step, because it requires so much more review overhead. The others I think we're all more on board with so those are easier to ship. Does that make sense?

**Comment by zljubisic** [comment]
Linux terminator 

image

uses this kind of logic:
every pane can belong to a group that can have arbitrarily name. Than you can press alt+a in order to type in all panes no mether how panes are grouped. Or you can press alt+g in order to type in all panes that are in the same group as the current pane. Or you can press alt+o in order to type only in the current pane.

You can set the group with menu accessible from each pane.

image

For example at the picture bellow, you can see that three panes are in the same group "sboddy".

image

**Comment by zadjii-msft** [comment]
Thanks for the feedback @zljubisic. I've moved that comment over to #9365 where we're discussing more of the big picture ideas for broadcast input mode. Thanks!

**Comment by zadjii-msft** [comment]
> Another option related to this one is an ability to define layout with panes and initial commands.

You can already do that today with the [`startupActions`](https://docs.microsoft.com/en-us/windows/terminal/customize-settings/startup#startup-actions) setting

**Comment by zljubisic** [comment]
Yes but I am talking about predefined layout, so I can invoke the layout, and I will get shell prompt on all severs.
I don't want to recreate layout from very beginning.

**Comment by zadjii-msft** [comment]
So for reference, I like to do like the following each time I boot up the Terminal in the morning:

```json
{ "command": { "action": "wt", "commandline": "new-tab --title OpenConsole cmd.exe /k #work 15 ; split-pane -s .30 --title OpenConsole cmd.exe /k #work 15 ; split-pane  -s .25 -H cmd.exe /k media ; new-tab --title \"Symbols Script\" powershell dev\\symbols.ps1 ; new-tab -p \"Ubuntu 18.04\" ; new-tab -p \"microsoft/Terminal\" ; sp -V -p \"microsoft/Terminal\" ; sp -H -p \"microsoft/Terminal\" ; focus-tab -t 0" }, "name": "Good Morning" },

```

(you could probably make it shorter by replacing `new-tab`/`split-pane` with `nt`/`sp`, respectively).

I add that to the `keybindings`/`actions`. This creates a new command in the [Command Palette](https://docs.microsoft.com/en-us/windows/terminal/command-palette) named "Good Morning". That opens up a few tabs & panes, running various build environments. I like having it in a command rather than `startupActions`, because I only really want this in _one_ terminal window, not _every single one I launch_. It's personal taste. 

You could repeat this for multiple different "session"s if you wanted. That way you could have layouts pre-defined for various different dev environments.

**Comment by zljubisic** [comment]
Nice. Give to me some time in order to experiment with it a little bit.
Thanks for sharing your setup.

**Comment by zljubisic** [comment]
@zadjii-msft Can you please help me with size parameter. There is not much on https://docs.microsoft.com/en-us/windows/terminal/command-line-arguments?tabs=windows.

If I want to divide the terminal in let's say 3 equally sized vertical panes, or four panes in 2x2 format, how to do that?

**Comment by zadjii-msft** [comment]
broadcast-resurrection-001

**Pull request closed without merging**